ft-logic/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+#![no_std]
+use ft_logic_io::*;
+use ft_storage_io::*;
+use gstd::{debug, msg, prelude::*, prog::ProgramGenerator, ActorId};
+mod instruction;
+use instruction::*;
+use primitive_types::H256;
+
+const GAS_STORAGE_CREATION: u64 = 3_000_000_000;
+
+#[derive(Default)]
+struct FTLogic {
+    admin: ActorId,
+    ftoken_id: ActorId,
+    transaction_status: BTreeMap<H256, TransactionStatus>,
+    instructions: BTreeMap<H256, (Instruction, Instruction)>,
+    storage_code_hash: H256,
+    id_to_storage: BTreeMap<String, ActorId>,
+}
+
+static mut FT_LOGIC: Option<FTLogic> = None;
+
+pub enum TransactionStatus {
+    InProgress,
+    Success,
+    Failure,
+}
+
+impl FTLogic {
+    /// The message received from the main contract.
+    ///
+    /// Arguments:
+    /// * `transaction_hash`: the hash associated with that transaction;
+    /// * `account`: the account that sent the message to the main contract;
+    /// * `action`: the message payload.
+    async fn message(&mut self, transaction_hash: H256, account: &ActorId, action: &Action) {
```

Comment:
No, because Action is transferred from account who calls the `main` contract and then `main` contract adds `msg::source()`to the message and redirects to the `logic` contract